### ISSUE TITLE
Getting started

### ISSUE THREAD

**Issue by cboettig** [cboettig/rappl#1]
@ironholds Thanks, here's the repo I mentioned.

My main problem is that I don't really know what I'm doing.  The code I want to wrap, [APPL](http://bigbird.comp.nus.edu.sg/pmwiki/farm/appl/) is a handful of C++ programs that are usually just downloaded, compiled, and then run with a suite a command-line options.  Here's a copy of the original APPL README (from linux sources, they distribute separate sources for mac): https://github.com/cboettig/rappl/blob/master/inst/README.txt

That README shows the typical command line usage, e.g.

```
pomdpsol ../examples/POMDPX/Tiger.pomdpx
```

I'd like to have an R function in place of the `pomdpsol` program that does the same thing (e.g. takes the file path as an argument, along with other optional arguments that `pomdpsol` permits on the command line), like this:

```r
pomdpsol("../examples/POMDPX/Tiger.pomdpx")
```

and so forth.  (Later one might imagine replacing the input and outputs, which are currently all file paths, into R objects, but that's straight forward R stuff I can handle).  

Wrapping sounds like it should be straight forward (either with a little RCpp wrapper or just `.Call` directly), but I'm getting stuck in a bunch of places.  

My understanding is that I'd want to replace the Makefile shipped with the sourcecode with a Makevars file which includes the necessary CFLAGS etc but no targets, letting R's default handle that.  But I'm not quite sure how to go about that and still make sure everything links up appropriately.

Also, I'm not quite sure where to best to make the connection.  The APPL source includes a makefile which generates ~5 executable scripts which I'd like to replace with R function wrappers.  This means that each is has some source code with a `int main(int argc, char **argv)` function, which I suppose is the thing I want to wrap but makes for a bit of a clumsy interface since the argument is just a character string instead of usual objects.  Alternatively I could imagine just bypassing `main` routine and wrapping the various subroutines called by `main`, but a look at the source code suggests this isn't trivial since there's a lot of logic in some of the `main` routines themselves (e.g. for pomdpsol: https://github.com/cboettig/rappl/blob/master/src/OfflineSolver/solver.cpp#L435 ).    

Does that make sense?  Any ideas how to get going here?

**Comment by Ironholds** [comment]
Gotcha; makes sense! Wrapping main() is probably the direction I'd go in, but using 'wrap' very broadly (i.e. cutting out the normal arguments and switching them over to R-like objects).  I mean, a char from an R string is just r_string.c_str() if you ask for a std::string rather than a CharacterVector.

I'll probably have some time to work on this late next week, if that works?

**Comment by cboettig** [comment]
Sure, would be awesome if you took a whack at it when you get a chance.  

Yup, wrapping `main` sounds like the way to go.  Best to have a solution that involves as shallow surgery to the underlying src code as possible, since that would also be easiest to maintain in the event of any newer release of the original source code.

**Comment by Ironholds** [comment]
So it looks like wrapping main() may not be possible because it's not evaluating the arguments _in_ main, it's evaluating them further down the pipe, and turning things R-side into an arglist is...well, it has 'arg' in it for a reason. So I may have to yank out and up a bit of logic: we'll see.

**Comment by cboettig** [comment]
yeah, interesting.  if it's simpler to get things going, it would be fine to pass the raw string arg list, flags and all to the R function, e.g. 

```r
pomdpsol("../examples/POMDPX/Tiger.pomdpx  --precision 25")
```

instead of the more natural construction like:

```r
pomdpsol(file = "../examples/POMDPX/Tiger.pomdpx",  precision = 25)
```

(Obviously we could construct the latter from the former by doing some string parsing on the R side anyway; not ideal but a lot better than a `system()` call to an externally installed utility...)

does that sound reasonable?

**Comment by Ironholds** [comment]
OOOH, that's a great idea. Much easier than my current "write out all the arguments and cuss" approach. an R wrapper with functions and some checks and a sprintf() call to tie it together!

**Comment by cboettig** [comment]
@Ironholds realized I hadn't granted you commit access earlier.  If you push what you've got I can take a whack at some of the wrapper side here, just can't figure out how to wrap the `main()` routines & still get the package to build in the first place...

**Comment by Ironholds** [comment]
Hey Carl,

So, my recommendation would be that you won't be able to :/. main() is a prohibited function in R compiled code, and I also see some assert and error statements in there. So my recommendation would be to take the body of the main functions, throw them in a dedicated Rcpp function, and go from there, combing through the body of the code all the way down to check for said assert and exit calls. Damn standalone libs :/.

I would love to help out but, see http://ironholds.org/blog/down-and-out-in-statistical-computing/
